### ISSUE TITLE
[SPIRV] Issue with non-uniform indexing

### ISSUE THREAD

**Issue by pmichniewski** [microsoft/DirectXShaderCompiler#2436]
Hi,

I'm having an issue when trying to use non-uniform indexing. The generated shader behaves as if it only had dynamic indexing but not non-uniform. I modified Sascha Willems' example to illustrate this. I can upload my modifications somewhere if needed. Here is the sample code that triggers this issue:
```
sampler samp : register(s1);
Texture2D<float4> textures[] : register(t2);

struct VS_OUTPUT
{
	float2 inUV : TEXCOORD0;
	nointerpolation int inTexIndex : TEXCOORD1;
};

struct PS_OUTPUT
{
	float4 outFragColor : SV_Target0;
};

PS_OUTPUT main(VS_OUTPUT i) 
{
	PS_OUTPUT o;
	o.outFragColor = textures[NonUniformResourceIndex(i.inTexIndex)].Sample(samp, i.inUV);
	return o;
}
```

When I compare the resulting SPIRV to what the equivalent GLSL code produces, the only obvious difference is the lack of OpCopyObject on the index that is used to access the texture array. I'm also attaching the HLSL and GLSL sources and their disassembly and a screenshot of this issue.

[descriptorindexing.glsl.txt](https://github.com/microsoft/DirectXShaderCompiler/files/3565594/descriptorindexing.glsl.txt)
[descriptorindexing.hlsl.txt](https://github.com/microsoft/DirectXShaderCompiler/files/3565595/descriptorindexing.hlsl.txt)
[descriptorindexing_disasm_glsl.txt](https://github.com/microsoft/DirectXShaderCompiler/files/3565598/descriptorindexing_disasm_glsl.txt)
[descriptorindexing_disasm_hlsl.txt](https://github.com/microsoft/DirectXShaderCompiler/files/3565599/descriptorindexing_disasm_hlsl.txt)
![indexing](https://user-images.githubusercontent.com/1138999/64108157-6e333800-cd7c-11e9-920c-9743eee10835.jpg)

**Comment by pmichniewski** [comment]
While looking around I found the following commit to glslang which seems to indicate that the lack of OpCopyObject is in fact the issue here: https://github.com/KhronosGroup/glslang/pull/1762.

**Comment by pmichniewski** [comment]
After more poking around, I found out that DXC actually generates an additional OpLoad for the index which is decorated by nonuniformEXT but it gets elided by the optimizer. If I compile the shader with -Od it works fine.

**Comment by pmichniewski** [comment]
I managed to hack around this particular issue by recursively applying setNonUniform for every instruction pointed to by the instruction marked with NonUniformResourceIndex() (basically for each load/access chain, etc. in the chain). This solved this simple test case but it still didn't fix my original problem which can be described by a more complicated test case:

```
sampler samp : register(s1);
struct uvMapping {
	float2 uv;
	uint index;
};
StructuredBuffer<uvMapping> mapping : register(t2);
Texture2D<float4> textures[] : register(t4);

struct VS_OUTPUT
{
	float2 inUV : TEXCOORD0;
	nointerpolation int inTexIndex : TEXCOORD1;
};

struct PS_OUTPUT
{
	float4 outFragColor : SV_Target0;
};

uint mappingCount;

PS_OUTPUT main(VS_OUTPUT inp) 
{
	PS_OUTPUT o;
	uint index = 0;
	[loop]
	for (uint i = 0; i < mappingCount; ++i)
	{
		if (all(inp.inUV <= mapping[i].uv)) {
			index = mapping[i].index;
			break;
		}
	}
	o.outFragColor = textures[NonUniformResourceIndex(index)].Sample(samp, inp.inUV);
	return o;
}
```

This still fails with similar artifacts:
![indexing2](https://user-images.githubusercontent.com/1138999/64179681-ddd02280-ce63-11e9-8686-83c40946bff1.jpg)

Basically, this issue isn't caused by the missing OpCopyObject but rather by missing nonuniformEXT decorations for each instruction accessing non-uniform data.

**Comment by ehsannas** [comment]
Thanks for reporting, @pmichniewski. And thanks for the detailed investigation. We'll take a closer look.

**Comment by pmichniewski** [comment]
I managed to create a workaround that fixes all my issues. It didn't work properly until I updated to the latest DXC (spirv-tools/opt was recently modified to better handle nonuniformEXT decorations - https://github.com/KhronosGroup/SPIRV-Tools/pull/2716/commits/63787b61fae00292288f47afad71f45688a5dfed). I'm attaching my workaround. It might point you towards a more proper solution.

[nonuniform_workaround.txt](https://github.com/microsoft/DirectXShaderCompiler/files/3573995/nonuniform_workaround.txt)

**Comment by ehsannas** [comment]
Thanks for all the information @pmichniewski . 

* I read your workaround. It is not clear to me why the decoration should be applied to a copy of the index, rather than the index itself (**I don't understand why glslang does that either**)? i.e. I don't understand why OpCopyObject is needed. Can we avoid it?
* Do you know if using the recent SPIRV-Tools alone fixes the issue? (I expect DXC code as it is today with the updated SPIRV-Tools should work).
* If not, does your work around minus the OpCopyObject work?

**Comment by pmichniewski** [comment]
Hi @ehsannas,

- I'm also not 100% sure why it has to be applied to the copy but from my observations, it didn't work without the copy. As far as I understand SPIRV-Tools specifically looks for a copy marked with nonuniformEXT - https://github.com/KhronosGroup/SPIRV-Tools/commit/63787b61fae00292288f47afad71f45688a5dfed#diff-c54f19e7fe00cd164d8304d06d26fb0aR93 - I think that's just a consequence of how glslang handles things (it does a copy because vertex input constructors don't allow for nonuniformEXT decorations). That's why I call my "fix" a workaround - it just happens to work.
- No, updating SPIRV-Tools alone does not fix the issue.
- No, it didn't work without the copy because of the reason I mentioned earlier.

Basically from what I understand from the specs, not only the sample itself but also the index and every step needed for getting the index (reading from vertex inputs, buffers, etc.) should be marked with the decoration.

**Comment by ehsannas** [comment]
@pmichniewski I have read the spec for SPIR-V and Vulkan one more time. From the Vulkan spec:

> If an instruction loads from or stores to a resource (including atomics and image instructions) and the resource descriptor being accessed is not dynamically uniform, then the operand corresponding to that resource (e.g. the pointer or sampled image operand) must be decorated with NonUniformEXT.

I also confirmed my understanding via https://github.com/KhronosGroup/glslang/issues/1760#issuecomment-530083454.

Baased on the discussion on the glslang issue and the reading of the spec, I believe the current implementation of DXC is correct: We start by decorating the loaded value of the `NonUniformResourceIndex` operand, and we propagate the decoration **forward** until we reach an access instruction. And based on Jeff's https://github.com/KhronosGroup/glslang/issues/1760#issuecomment-530094882:
> what really matters is only whether the operand of the memory access instruction is decorated as nonUniform

So I think the current DXC implementation is correct.

----

> the index and every step needed for getting the index

The **backward** propagation that you described above should not be needed based on the above discussion. So, this might be a case of a driver bug. Do you have the latest drivers? Can you try another vendor's GPU if you have access to one?

**Comment by pmichniewski** [comment]
@ehsannas I checked this on latest AMD drivers. I'll also check this on an nVidia card once I get to work tomorrow.

I still think DXC is not behaving 100% correct here since:

- it does not decorate reading from the variable to be used as an index (it does decorate an OpLoad but this OpLoad gets elided by the optimizer) - this I will actually have to check again since I tried this before the change to the optimizer was introduced
-  it does not decorate reading a value to be used as an index from a structured buffer - this definitely still happens even after the update to the optimizer

I think the problem stems from the fact that you only apply the decoration to the index itself, but the index might come from either a structured buffer, an input variable, or in my case a combination of both with some calculation in between. Note that my test case is a very simplified example - the original shader in which this happens is much more complicated and my workaround also fixes it. Moreover, in the test case, the if statements get replaced by OpPhi but in the original shader OpCompositeExtract are used.

**Comment by ehsannas** [comment]
* If the optimizer is eliminating the an OpLoad that has NonUniformEXT, we should make sure the optimizer is fixed.
* Based on your explanation I think I was able to come up with the code pattern(s) that can produce incorrect SPIR-V:

I believe we generate **correct** SPIR-V for the following cases:
```hlsl
Texture2D           gTextures[32];
SamplerState        gSamplers[];
RWStructuredBuffer<uint> sb;
// ...
float4 a = gTextures[NonUniformResourceIndex(index)].Sample(gSamplers[0], loc);
float4 b = gTextures[sb[NonUniformResourceIndex(index)]].Sample(gSamplers[0], loc);
float4 c = gTextures[NonUniformResourceIndex(sb[index])].Sample(gSamplers[0], loc);
uint d = sb[NonUniformResourceIndex(index)];
```

However, if there are calculations performed on the operand of `NonUniformResourceIndex`, it looks like we will be missing the decoration going all the way back to the loading of the variables. For example:
```hlsl
Texture2D           gTextures[32];
SamplerState        gSamplers[];
// ...
float4 a = gTextures[NonUniformResourceIndex(index + 1 + 2)].Sample(gSamplers[0], loc);
```

here we generate:
```
               OpDecorate %58 NonUniformEXT
               OpDecorate %60 NonUniformEXT
               OpDecorate %61 NonUniformEXT
               OpDecorate %67 NonUniformEXT
; ...
         %56 = OpLoad %uint %index
         %57 = OpIAdd %uint %56 %uint_1
         %58 = OpIAdd %uint %57 %uint_2
         %60 = OpAccessChain %_ptr_UniformConstant_type_2d_image %gTextures %58
         %61 = OpLoad %type_2d_image %60
         %63 = OpAccessChain %_ptr_UniformConstant_type_sampler %gSamplers %int_0
         %64 = OpLoad %type_sampler %63
         %65 = OpLoad %v2float %loc
         %67 = OpSampledImage %type_sampled_image %61 %64
         %68 = OpImageSampleImplicitLod %v4float %67 %65 None
               OpStore %v1 %68
```

We have applied the decoration to the result of the addition (`%58`), but not traced it back to loading of the index (`%56`). This is because the current code does not trace back at all; it only propagates forward.
In other words, `NonUniformResourceIndex(index) + 1` would give correct results whereas `NonUniformResourceIndex(index + 1)` would not.

**Comment by ehsannas** [comment]
Also as was pointed out, for
```
 gTextures[NonUniformResourceIndex(index + 1 + 2 + sb[index])].Sample(gSamplers[0], loc);
```
`sb[index]` generates two OpLoad instructions for which the first one is missing the decoration.

**Comment by pmichniewski** [comment]
Yes, I think this is a great summary of the issue. You're much better at explaining this than I am :)
I will get back to you on the first issue tomorrow as it might already be fixed - I didn't test this simple case after the optimizer was updated yet.

**Comment by ehsannas** [comment]
Thanks!

**Comment by ehsannas** [comment]
@pmichniewski I almost have an implementation ready...
BUT It suddenly occurred to me that I may have thought about applying the decoration to the wrong Loads. For example, for:

```
gTextures[NonUniformResourceIndex(index + 1 + sb[index])].Sample(gSamplers[0], loc);
```
The `OpLoad`s that are related to loading the `index` and loading `sb[index]` are **not** NonUniform. It is the `OpLoad` related to indexing into `gTextures` that is NonUniform.

Take the following GLSL/glslang as an example:
http://shader-playground.timjones.io/3eec3f994ff5c6771927018006e698c0

```
              17:     14(int) Load 16(rIndex)
              19:     14(int) IAdd 17 18
              21:     14(int) IAdd 19 20
              22:     14(int) CopyObject 21
              24:     23(ptr) AccessChain 13(data) 22
              25:          10 Load 24
```
The decoration is only applied to `22` and `25` (and theoretically if copyObject was not done, it would only be applied to `25`)... Loading `rIndex` (`17`) does **not** get the decoration.
